LuaRules/Gadgets/unit_terraform.lua

```diff
@@ -3619,64 +3620,34 @@ local function deregisterStructure(unitID)
 	
 end
 
-local dirtbagPosX =
-	          { -8, 0, 8,
-	        -16,-8, 0, 8, 16,
-	    -24,-16,-8, 0, 8, 16, 24,
-	-32,-24,-16,-8, 0, 8, 16, 24, 32,
-	-32,-24,-16,-8, 0, 8, 16, 24, 32,
-	-32,-24,-16,-8, 0, 8, 16, 24, 32,
-	    -24,-16,-8, 0, 8, 16, 24,
-	        -16,-8, 0, 8, 16,
-	            -8, 0, 8}
-
-local dirtbagPosZ =
-	           {-32,-32,-32,
-	        -24,-24,-24,-24,-24,
-	    -16,-16,-16,-16,-16,-16,-16,
-	 -8, -8, -8, -8, -8, -8, -8, -8, -8,
-	  0,  0,  0,  0,  0,  0,  0,  0,  0,
-	  8,  8,  8,  8,  8,  8,  8,  8,  8,
-	      16, 16, 16, 16, 16, 16, 16,
-	          24, 24, 24, 24, 24,
-	              32, 32, 32}
-
-local dirtbagPosY =
-	            {3 , 4 , 3 ,
-	         3 , 5 , 9 , 5 , 3 ,
-	     3 , 6 , 25, 27, 25, 6 , 3 ,
-	 3 , 5 , 25, 29, 31, 29, 25, 5 , 3 ,
-	 4 , 9 , 27, 31, 32, 31, 27, 9 , 4 ,
-	 3 , 5 , 25, 29, 31, 29, 25, 5 , 3 ,
-	     3 , 6 , 25, 27, 25, 6 , 3 ,
-	          3, 5 , 9 , 5 , 3 ,
-	             3 , 4 , 3 }
-
 function gadget:UnitDestroyed(unitID, unitDefID)
 
-	if (unitDefID == shieldscoutDefID) then
-		local  _,_,_,_,build = spGetUnitHealth(unitID)
-		if build == 1 then
-			local ux, uy, uz  = spGetUnitPosition(unitID)
+	if (terraformOnUnitDestroyed[unitDefID]) then
+		local  _,_,_,_,buildProgress = spGetUnitHealth(unitID)
+		if buildProgress == 1 then
+			local config = terraformOnUnitDestroyed[unitDefID]
```

Comment:
Could avoid indexing the table twice:
```suggestion
	local config = terraformOnUnitDestroyed[unitDefID]
	if config then
		local  _,_,_,_,buildProgress = spGetUnitHealth(unitID)
		if buildProgress == 1 then
```

Reply:
But then you would create an extra variable and perform assignment for all units, instead of extra indexing for Dirtbags only.
No idea what is better tbh.

Reply:
I checked the "compiler" output (`$ luac -l -l -p indexing.lua`). This chunk:
```lua
local tab
local function single(index)
             local x = tab[index]
             if x then
             end
end

local function double(index)
             if tab[index] then
                          local x = tab[index]
             end
end
```
has this output:
```
function <indexing.lua:2,6> (5 instructions, 20 bytes at 0x19697c0)
1 param, 2 slots, 1 upvalue, 2 locals, 0 constants, 0 functions
	1	[3]	GETUPVAL 	1 0	; tab
	2	[3]	GETTABLE 	1 1 0
	3	[4]	TEST     	1 0 0
	4	[4]	JMP      	0	; to 5
	5	[6]	RETURN   	0 1
constants (0) for 0x19697c0:
locals (2) for 0x19697c0:
	0	index	1	5
	1	x	3	5
upvalues (1) for 0x19697c0:
	0	tab

function <indexing.lua:8,12> (7 instructions, 28 bytes at 0x1969710)
1 param, 2 slots, 1 upvalue, 2 locals, 0 constants, 0 functions
	1	[9]	GETUPVAL 	1 0	; tab
	2	[9]	GETTABLE 	1 1 0
	3	[9]	TEST     	1 0 0
	4	[9]	JMP      	2	; to 7
	5	[10]	GETUPVAL 	1 0	; tab
	6	[10]	GETTABLE 	1 1 0
	7	[12]	RETURN   	0 1
constants (0) for 0x1969710:
locals (2) for 0x1969710:
	0	index	1	7
	1	x	7	7
upvalues (1) for 0x1969710:
	0	tab
```
So my way is strictly superior. The temporary value from the first indexing has to be put on the stack anyway, so there is no extra assignment (it even gets put in the same stack slot, so no memory either). Indexing twice just results in an additional `GETTABLE` op (and `GETUPVAL` since the table is not local). It doesn't matter that the `if` is empty, putting things inside is symmetrical.

Reply:
Thanks for doing the research!
Thats certainly some stuff worth to know. I will remember this for future code.
As for this case, I will fix it in next PM.

Reply:
Already taken care of 737a07c23bdcaa9282e17f8365b86fd124614d80